File: 벽부수고이동하기.cpp
#ifdef _MSC_VER
#define _CRT_SECURE_NO_WARNINGS
#endif
#include <iostream>
#include <queue>
using namespace std;

struct Point {
	Point() {};
	Point(int x, int y, int distance, int wall, int value, bool visited) :X(x), Y(y), DISTANCE(distance), WALL(wall), VALUE(value), VISITED(visited) {}
	int X;
	int Y;
	int DISTANCE;
	int WALL;
	int VALUE;
	bool VISITED;
};

int N, M;
Point area[1000][1000];
int dir[4][2] = { {1, 0},{-1,0},{0, 1},{0, -1} };

int bfs(Point point) {
	queue<Point> q;
	q.push(point);

	while (!q.empty()) {
		Point p = q.front();
		q.pop();

		printf("%d %d %d %d %d \n", p.X,
			p.Y,
			p.WALL,
			p.DISTANCE, p.VALUE);

		if (p.X == M - 1 && p.Y == N - 1)
			break;
		for (int i = 0; i < 4; i++) {
			if (p.X + dir[i][0] > -1 && p.X + dir[i][0] < M && p.Y + dir[i][1] > -1 && p.Y + dir[i][1] < N && p.WALL < 2) {
				if (area[p.X + dir[i][0]][p.Y + dir[i][1]].VALUE == 1)
					area[p.X + dir[i][0]][p.Y + dir[i][1]].WALL += 1;
				area[p.X + dir[i][0]][p.Y + dir[i][1]].DISTANCE += 1;
				q.push(area[p.X + dir[i][0]][p.Y + dir[i][1]]);
				
			}
		}
	}

	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			printf("%d %d %d %d %d \n", area[i][j].X,
				area[i][j].Y,
				area[i][j].WALL,
				area[i][j].DISTANCE, area[i][j].VALUE);

		}
	}
	printf("\n");

	if (area[N - 1][M - 1].DISTANCE == 0)
		return -1;
	return area[N - 1][M - 1].DISTANCE;
}

int main() {
	scanf("%d %d", &N, &M);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < M; j++) {
			scanf("%1d", &area[i][j].VALUE);
			area[i][j].X = j;
			area[i][j].Y = i;
			area[i][j].WALL = 0;
			area[i][j].DISTANCE = 0;
			area[i][j].VISITED = false;
		}
	}
	bfs(area[0][0]);
	
}

File: 벽 부수고 이동하기.py
import queue
import copy

answer = False

# point 각각의 경로에 대해서 방문여부를 기록한다.

def bfs(q, str, n, m):
    dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    global answer

    _q = queue.Queue()
    while q.empty() == False:
        p = q.get()
        for d in dir:
            if p[0] + d[0] == n - 1 and p[1] + d[1] == m - 1:
                answer = True
            if p[0] + d[0] > -1 and p[0] + d[0] < n and p[1] + d[1] > -1 and p[1] + d[1] < m:
                if p[3][p[0]+d[0]][p[1]+d[1]] == 0:
                    p[3][p[0]+d[0]][p[1]+d[1]] = 1
                    if str[p[0] + d[0]][p[1] + d[1]] == 0:
                        _q.put([p[0]+d[0], p[1]+d[1], p[2], p[3]])
                    if str[p[0] + d[0]][p[1] + d[1]] == 1:
                        if p[2] == 0:
                            _q.put([p[0]+d[0], p[1]+d[1], p[2] + 1, p[3]])
    return _q


a = input().split(' ')
n, m = int(a[0]), int(a[1])

str = []
for i in range(0, n):
    str.append(list(input()))

for i in range(0, n):
    for j in range(0, m):
        str[i][j] = int(str[i][j])

temp = [[0] * m for i in range(0, n)]

q = queue.Queue()
q.put([0, 0, 0, temp])

min = 1

while True:
    if answer:
        print(min)
        break
    else:
        q = bfs(q, str, n, m)

        if q.empty() == True:
            print(-1)
            break
        min+=1
